### PR TITLE
[GTKUI] Fix #2802: GTKUI classic mode shutdown procedure is broken

### PR DESCRIPTION
**Pull request by bendikro**
As explained in ticket, current shutdown routine in GTKUI is broken: http://dev.deluge-torrent.org/ticket/2802

I think this fixes the issues, but only tested on Ubuntu 15.10

### CODE REVIEW

File: deluge/ui/gtkui/gtkui.py [hunk 1]
```diff
@@ -239,35 +240,51 @@ def nsapp_open_file(osxapp, filename):
...
-    def shutdown(self, *args, **kwargs):
-        log.debug("gtkui shutting down..")
-
-        component.stop()
 
         # Process any pending gtk events since the mainloop has been quit
         while gtk.events_pending():
-            gtk.main_iteration(0)
+            gtk.main_iteration_do(block=False)
+            time.sleep(0.0001)
```
Comment (cas--) [inline deluge/ui/gtkui/gtkui.py]: Is this very short sleep required?

File: deluge/ui/gtkui/gtkui.py [hunk 2]
```diff
@@ -239,35 +240,51 @@ def nsapp_open_file(osxapp, filename):
 
         # Twisted catches signals to terminate, so have it call a pre_shutdown method.
         reactor.addSystemEventTrigger("before", "shutdown", self.pre_shutdown)
```
Comment (cas--) [inline deluge/ui/gtkui/gtkui.py]: Does this get triggered now?

File: deluge/core/torrentmanager.py [hunk 3]
```diff
@@ -693,7 +693,7 @@ def _save_state(self):
                 os.fsync(_file.fileno())
         except (OSError, cPickle.PicklingError) as ex:
             log.error("Unable to save %s: %s", filename, ex)
-            return
+            return False
```
Comment (cas--) [inline deluge/core/torrentmanager.py]: Are these torrentmanager changes related to this PR?

File: deluge/ui/gtkui/gtkui.py [hunk 4]
```diff
@@ -239,35 +240,51 @@ def nsapp_open_file(osxapp, filename):
...
         """Modal dialogs can prevent the application exiting so destroy mainwindow"""
+        # Make sure the config is saved.
+        self.config.save()
         # Ensure columns state is saved
         self.torrentview.save_state()
         self.mainwindow.window.destroy()
 
+    @defer.inlineCallbacks
+    def close(self):
+        if self.closing:
```
Comment (cas--) [inline deluge/ui/gtkui/gtkui.py]: Probably better with `if not self.closing` and drop the `return`

File: deluge/core/torrentmanager.py (hunk 3 above)
Comment (bendikro) [inline deluge/core/torrentmanager.py]: Umm...no
I checked the return value during testing, but it's not needed for the PR. Would be reasonable to return a value or success/failure? Not strictly used for anything at the moment tho.

File: deluge/ui/gtkui/gtkui.py (hunk 1 above)
Comment (bendikro) [inline deluge/ui/gtkui/gtkui.py]: Not sure, just found this: http://stackoverflow.com/questions/2844902/python-four-steps-setup-with-progressbars
I was going to ask whether this code is even necessary at all...?

File: deluge/ui/gtkui/gtkui.py (hunk 2 above)
Comment (bendikro) [inline deluge/ui/gtkui/gtkui.py]: I've combined the two functions.

File: deluge/ui/gtkui/gtkui.py (hunk 4 above)
Comment (bendikro) [inline deluge/ui/gtkui/gtkui.py]: I'd say it's best to avoid indentations of possible?

File: deluge/ui/gtkui/gtkui.py (hunk 1 above)
Comment (cas--) [inline deluge/ui/gtkui/gtkui.py]: I'm sure I looked into it before but forgotten the original reason. As this is on shutdown the delay should be unneeded.

One of the original sources for this code is probably the pygtk faq: http://faq.pygtk.org/index.py?req=all

File: deluge/core/torrentmanager.py (hunk 3 above)
Comment (cas--) [inline deluge/core/torrentmanager.py]: I think without a concrete idea of return value, it's best left without so it's relatively obvious

File: deluge/ui/gtkui/gtkui.py (hunk 1 above)
Comment (bendikro) [inline deluge/ui/gtkui/gtkui.py]: The comment says `# Process any pending gtk events since the mainloop has been quit`, but what events would this be? When performing a long running task the GUI is unable to update making it seem to freeze. In which case this code would be useful.

Maybe it was added to try to fix issues with the GTKUI not properly closing in standalone mode when closing with CTRL-C?
If that is the case, it's not needed, as it does not fix that problem, and the issue is fixed by this PR anyways.

Comment (cas--) [inline deluge/ui/gtkui/gtkui.py]: The changes were made in 790a10e3, d0fb657adaa, a3636ccd and 848b14605eb3b but there are few details of the whys...

File: deluge/ui/gtkui/gtkui.py [hunk 5]
```diff
@@ -238,36 +239,52 @@ def nsapp_open_file(osxapp, filename):
...
-    def pre_shutdown(self, *args, **kwargs):
-        """Modal dialogs can prevent the application exiting so destroy mainwindow"""
         # Ensure columns state is saved
         self.torrentview.save_state()
+        # Shut down components
+        yield self.shutdown()
+
+        # Modal dialogs can prevent the application exiting so destroy mainwindow
+        # Must do this here to avoid hang when closing with SIGINT (CTRL-C)
         self.mainwindow.window.destroy()
```
Comment (cas--) [inline deluge/ui/gtkui/gtkui.py]: This has to be done before shutdown as the comment describes and is tied to `self.torrentview.save_state()`

File: deluge/ui/gtkui/gtkui.py [hunk 6]
```diff
@@ -238,36 +239,52 @@ def nsapp_open_file(osxapp, filename):
...
+        reactor.addSystemEventTrigger("before", "gtkui_close", self.close)
+
         reactor.callWhenRunning(self._on_reactor_start)
+        self.closing = False
+
+        def gtkui_sigint_handler(num, frame):
+            log.debug("SIGINT signal caught - firing event: 'gtkui_close'")
+            reactor.callLater(0, reactor.fireSystemEvent, 'gtkui_close')
+
+        signal.signal(signal.SIGINT, gtkui_sigint_handler)
```
Comment (cas--) [inline deluge/ui/gtkui/gtkui.py]: To me this seems like the exact same as if twisted catches the sigint before shutdown.

Why not simply `reactor.addSystemEventTrigger("before", "shutdown", self.close)`?

File: deluge/ui/gtkui/gtkui.py (hunk 1 above)
Comment (bendikro) [inline deluge/ui/gtkui/gtkui.py]: They do explain the purpose of the changes:

- `Fix #535 fix saving column widths`
- `Fix exception when quitting in classic mode`
- `Fix saving torrent state on fresh configs`

So if all these work now it should be ok to remove the code. I haven't tested if the column widths are saved. State should be properly saved, but ideally there would be unit tests for this.

When testing with print statements the save state function did run properly, because the whole component shutdown process is now ensured to be performed before the reactor is shutting down.

I have not tested on Windows, which probably would be a good idea.

File: deluge/ui/gtkui/gtkui.py (hunk 5 above)
Comment (bendikro) [inline deluge/ui/gtkui/gtkui.py]: It is done before reactor shutdown. Before the last rebase, pre_shutdown was actually run after the close function, but now the same code is run before.

File: deluge/ui/gtkui/gtkui.py (hunk 6 above)
Comment (bendikro) [inline deluge/ui/gtkui/gtkui.py]: Well, it doesn't work. GTKUI ends up hanging.

Comment (bendikro) [inline deluge/ui/gtkui/gtkui.py]: Umm...after rebase on latest develop SIGINT gives new stacktrace.

Comment (cas--) [inline deluge/ui/gtkui/gtkui.py]: I was actually going to test this myself after lunch but thanks for checking.

File: deluge/ui/gtkui/gtkui.py (hunk 1 above)
Comment (cas--) [inline deluge/ui/gtkui/gtkui.py]: Considering for a while Windows was excluded from running this code I think it'll be safe to remove :D

Comment (bendikro) [inline deluge/ui/gtkui/gtkui.py]: Good point. Removed the code.

**Comment by bendikro** [comment]
Are there any unresolved issue left?

**Comment by cas--** [comment]
Merged in develop: 47f14845ca
